### PR TITLE
add where() operator and change and() and or()

### PR DESCRIPTION
**Pull request by staltz**
**BREAKING CHANGE**

Related to: #138 and #139

### CODE REVIEW

**Comment by staltz** [comment]
cc @Barbarrosa @mixmix

File: README.md [hunk 1]
```diff
@@ -159,11 +165,13 @@ query(
...
-    or(
-      equal(seekAuthor, aliceId, { indexType: 'author' }),
-      equal(seekAuthor, bobId, { indexType: 'author' })
+  where(
+    and(
+      equal(seekType, 'contact', 'type')
+      or(
+        equal(seekAuthor, aliceId, { indexType: 'author' }),
+        equal(seekAuthor, bobId, { indexType: 'author' })
+      )
```
Comment (arj03) [inline README.md:174]: nice

File: operators.js [hunk 2]
```diff
@@ -263,45 +263,18 @@ function not(ops) {
...
-          }
-        : rhs[0]
-    if (ops) copyMeta(ops, res)
+  return { type: 'OR', data: args }
+}
+
+function where(nextOp) {
+  return (prevOp) => {
+    if (!nextOp) return prevOp
+    const res = prevOp.type ? { type: 'AND', data: [prevOp, nextOp] } : nextOp
```
Comment (arj03) [inline operators.js:276]: This is actually a bit easier to read than the others. If I read this correctly it is: support null, support normal case, support where outside chain.

File: test/operators.js [hunk 3]
```diff
@@ -167,11 +159,11 @@ prepareAndRunTest('and() ignores non-function arguments', dir, (t, db, raf) => {
...
 
-prepareAndRunTest('or() ignores non-function arguments', dir, (t, db, raf) => {
+prepareAndRunTest('where() takes only one arguments', dir, (t, db, raf) => {
   const queryTree = query(
     fromDB(db),
-    or(
-      null,
+    where(
+      equal(helpers.seekType, 'contact', { indexType: 'type', indexAll: true }),
       equal(helpers.seekType, 'post', { indexType: 'type', indexAll: true })
```
Comment (arj03) [inline test/operators.js:167]: in this case the second argument is ignored, right?

File: operators.js (hunk 2 above)
Comment (staltz) [reply operators.js:276]: Yes, overall this file had a net negative git diff. You guessed correctly, but the 3rd case is "support `where` when it's the first thing right after `fromDB(db)`".

File: test/operators.js (hunk 3 above)
Comment (arj03) [reply test/operators.js:167]: I wonder if it would be better to throw an error

File: operators.js (hunk 2 above)
Comment (arj03) [reply operators.js:276]: Yes the negative diff in the amount of code is really nice and a good sign that this is a better abstraction.

File: test/operators.js (hunk 3 above)
Comment (staltz) [reply test/operators.js:167]: Hmm, yeah we can make `where` throw.

**Review by arj03 — APPROVED** [review]

**Comment by staltz** [comment]
Before merging, I'll try this out in ssb-db2 and see if it all works.

**Comment by mixmix** [comment]
I've just read the readme. Really sweet seeing these changes in response to that thread.
I appreciate how much work changing things like this around can be

# 🌱🍄🐉

**Comment by staltz** [comment]
Thanks @mixmix 

Ok, @arj03 I tested ssb-db2 and it works. I'll merge and make a new version later. It  will have to be a major version.

**Comment by Barbarrosa** [comment]
@staltz @arj03 Thanks, this makes JITDB queries far more predictable.

**Pull request merged**